### PR TITLE
caddyhttp: Implement placeholders in template

### PR DESCRIPTION
**Pull request by bagbag**
Pull request for #3314

I don't think that I have to say anything about the code, as it is little and straight-forward.

I decided to go for an template-function, I haven't found a better name than `placeholder`, feel free to recommend another name. In my opinion, the braces should be omitted in the context of templates. 

I'm not sure what text to return if the placeholder does not exist. Copy the behaviour of the `env`-function or return the placeholder unchanged, as it is done for placeholders in Caddyfile. I decided to go for the first one.

I'm not sure whether `modules/caddyhttp/placeholdershorthands.go` is the correct directory for that file or not or if there is a better/preferred way in Go/this code-base to avoid defining the list of shorthands multiple times.

I'm not sure about the documentation. Is the addition in `modules/caddyhttp/templates/templates.go` all I need for it to appear in the docs?
How can I edit <https://caddyserver.com/docs/conventions> to add an exception for the braces?

#### How to test this pull request

Create the following files next to the caddy binary (or the working directory for the running process):

`Caddyfile`
```
http://localhost {
  file_server
  templates
}
```

`template.html`
```
{{placeholder "http.request.host"}} - {{placeholder "path"}}
```

Run caddy: `./caddy run --config Caddyfile --adapter caddyfile`
Navigate to the file: <http://localhost/template.html>

### CODE REVIEW

**Comment by francislavoie** [comment]
FYI this implementation conflicts with https://github.com/caddyserver/caddy/pull/3305

I don't think it's right to expose the shorthands to the templates yet. I think that should be a separate change later.

**Comment by bagbag** [comment]
Good to know, I reverted these changes. If this pull request is merged using squash, there shouldn't be a conflict I think(?). Otherwise I can rewrite the history.

**Comment by francislavoie** [comment]
Nope, you should be good! Thanks!

FYI this may take a bit of time to get merged since we're still in a feature freeze until v2.0 stable is released.

**Comment by francislavoie** [comment]
For bonus points, you could add some tests for this to `tplcontext_test.go` if you're up for it.

File: modules/caddyhttp/templates/templates.go [hunk 1]
```diff
@@ -64,6 +64,16 @@ func init() {
 // {{env "VAR_NAME"}}
 // ```
 //
+// ##### `placeholder`
+//
+// Gets an [placerholder variable](/docs/conventions#placeholders). Shorthands are not supported at the moment.
```
Comment (mholt) [inline modules/caddyhttp/templates/templates.go:69]:
```suggestion
// Gets an [placeholder variable](/docs/conventions#placeholders).
```

Shorthand placeholders are a Caddyfile-only concept, doesn't make sense to document them in the JSON doc.

File: modules/caddyhttp/templates/tplcontext.go [hunk 2]
```diff
@@ -164,6 +166,18 @@ func (c templateContext) executeTemplateInBuffer(tplName string, buf *bytes.Buff
...
+	repl := c.Req.Context().Value(caddy.ReplacerCtxKey).(*caddy.Replacer)
+
+	value, has := repl.Get(name)
+
+	if !has {
+		return ""
+	}
+
+	return fmt.Sprintf("%v", value)
+}
```
Comment (mholt) [inline modules/caddyhttp/templates/tplcontext.go:179]:
```suggestion
	repl := c.Req.Context().Value(caddy.ReplacerCtxKey).(*caddy.Replacer)
	value, _ := repl.GetString(name)
	return value
}
```

Might need to rebase with master -- I know this GetString() method was added recently but I can't recall when, maybe it's still in a PR... in any case, please try this out.

**Comment by mholt** [comment]
Any interest in finishing this? If not, that's fine, I can close the PR so it's not hanging over our heads :)

**Comment by bagbag** [comment]
Oh, I totally forgot that PR. Yes I will finish this.

**Comment by bagbag** [comment]
Okay, so I commited your suggestions, rebased to master, built it, tested it (it still works) and forced pushed. Anything else missing for a merge?

**Comment by mholt** [comment]
Great! At this point I'm just trying to decide if `placeholder` is the right/best function name for this. I'm mostly indifferent about this, but, at first glance it doesn't strike me quite the right way. I wonder if there's a better term.

**Comment by bagbag** [comment]
I had similar thoughts when I started this PR. My best argument was to use the same terminology everywhere, eg. at https://caddyserver.com/docs/conventions#placeholders

When I searched for that feature, my query was "caddy placeholder in templates", as this was the term I knew from caddyfile.

**Comment by mholt** [comment]
Fair enough, we can roll with it for now and then change it later if needed (templates are all kind of experimental anyway, as the docs say).

Thanks for working on this and seeing it through to completion!

**Pull request merged**